libtenzir/builtins/contexts/geoip.cpp

```diff
@@ -0,0 +1,496 @@
+        auto output = record{};
+        entry_data_list_it
+          = entry_data_list_to_record(entry_data_list_it, &status, output);
+        if (status != MMDB_SUCCESS) {
+          return caf::make_error(
+            ec::lookup_error, fmt::format("error looking up IP address '{}' in "
+                                          "GeoIP database: {}",
+                                          ip_string, MMDB_strerror(status)));
+        }
+        auto r = field_builder.record();
+        r.field("address", value);
+        r.field("context", output);
+        r.field("timestamp", std::chrono::system_clock::now());
+      } else {
+        field_builder.null();
+      }
+    }
+    return field_builder.finish();
+  }
+
+  /// Inspects the context.
+  auto show() const -> record override {
+    return record{{path_key, db_path_}};
+  }
+
+  /// Updates the context.
+  auto update(table_slice, context::parameter_map)
+    -> caf::expected<record> override {
+    return caf::make_error(ec::unimplemented,
+                           "geoip context can not be updated with events");
+  }
+
+  auto update(chunk_ptr, context::parameter_map)
+    -> caf::expected<record> override {
+    return caf::make_error(ec::unimplemented, "geoip context can not be "
+                                              "updated with bytes");
+  }
+
+  auto update(context::parameter_map parameters)
+    -> caf::expected<record> override {
+    if (parameters.contains(path_key) and parameters.at(path_key)) {
+      db_path_ = *parameters[path_key];
+    }
+    if (mmdb_) {
+      MMDB_close(&*mmdb_);
+    } else {
+      mmdb_ = MMDB_s{};
+    }
+    auto status = MMDB_open(db_path_.c_str(), MMDB_MODE_MMAP, &*mmdb_);
+    if (status != MMDB_SUCCESS) {
+      return caf::make_error(ec::filesystem_error,
+                             fmt::format("error opening IP database at path "
+                                         "'{}': {}",
+                                         db_path_, MMDB_strerror(status)));
+    }
+    return show();
+  }
+
+  auto save() const -> caf::expected<chunk_ptr> override {
+    // We save the context by formatting into a record of this format:
+    // {path: string}
+    auto builder = flatbuffers::FlatBufferBuilder{};
+    const auto key_key_offset = builder.CreateSharedString(path_key);
+    const auto key_value_offset = pack(builder, data{db_path_});
+    auto field_offsets
+      = std::vector<flatbuffers::Offset<fbs::data::RecordField>>{};
+    field_offsets.reserve(1);
+    const auto record_offset
+      = fbs::data::CreateRecordField(builder, key_key_offset, key_value_offset);
+    field_offsets.emplace_back(record_offset);
+    const auto value_offset
+      = fbs::data::CreateRecordDirect(builder, &field_offsets);
+    const auto data_offset
+      = fbs::CreateData(builder, fbs::data::Data::record, value_offset.Union());
+    fbs::FinishDataBuffer(builder, data_offset);
+    return chunk::make(builder.Release());
+  }
```

Comment:
This serialization function is not easily adaptable for future fields. For the lookup table context, the format works because all we need to store is one (potentially heterogeneous) list of records. But for the geoip context, I think there is little reason to type-erase the state into a record to then persist that in a FlatBuffers file. Can we instead define a new FlatBuffers table for this context that (for now) just contains the path as a field?

Reply:
I simplified de-/serialization in https://github.com/tenzir/tenzir/pull/3731/commits/fab08465a199cc9ee61a9ea8309f31ecd54e17c6

Reply:
But this is still a single data inside a FlatBuffers table instead of a FlatBuffers table that is extensible. How do I add a field to what you proposed?

Reply:
Alright, I get behind the idea now - should the new `.fbs` file containing the `geoip-context` table be part of the `context` plugin in `tenzir-plugins`?

Reply:
I created the Flatbuffer table in https://github.com/tenzir/tenzir/pull/3731/commits/ef265b9ca3acedc3d8e38af451fc6cbbe2e5b251